File: algorithms/cpp/322.零钱兑换.cpp
#include <iostream>
#include <vector>

using namespace std;
/*
 * @lc app=leetcode.cn id=322 lang=cpp
 *
 * [322] 零钱兑换
 */

// @lc code=start
class Solution {
 public:
  /**
   * 状态数组: opt[i] 是 凑成总金额为 i 所需的最少硬币个数
   * 状态转移方程: dp[i]=min(dp[i - k]) - [k in coins]
   */
  int coinChange(vector<int>& coins, int amount) {
    vector<int> dp(amount + 1, amount + 1);
    dp[0] = 0;
    for (int i = 1; i < amount + 1; i++) {
      for (auto coin : coins) {
        if (i >= coin) dp[i] = min(dp[i], dp[i - coin] + 1);
      }
    }
    return dp[amount] > amount ? -1 : dp[amount];
  }
};
// @lc code=end

File: algorithms/golang/322.零钱兑换.go
/*
 * @lc app=leetcode.cn id=322 lang=golang
 *
 * [322] 零钱兑换
 */

// @lc code=start
/**
 * 类似爬楼梯， 每次可爬的步数为 coins中的数
 * 设 opt[i] 为凑够 i元所需最小的硬币数, k ∈ [k in coins]
 * 对于 opt[i]，opt[i]=min(opt[i],opt[i-k]+1)
 */
func coinChange(coins []int, amount int) int {
	opt := make([]int, amount+1)
	for i, _ := range opt {
		opt[i] = amount + 1
	}
	opt[0] = 0

	min := func(a, b int) int {
		if a < b {
			return a
		} else {
			return b
		}
	}
	for i := 1; i <= amount; i++ {
		for j := 0; j < len(coins); j++ {
			if i >= coins[j] {
				opt[i] = min(opt[i], opt[i-coins[j]]+1)
			}
		}
	}

	if opt[amount] > amount {
		return -1
	} else {
		return opt[amount]
	}
}

// @lc code=end
